### DISCUSSION TITLE
Improving consistency of 'unnumbered' links throughout the code base

### DISCUSSION THREAD

**Discussion by jbemmel** [ipspace/netlab#433] [General]
The way unnumbered links are created/treated has evolved, and has (imho) resulted in some complexity that could be removed.

The 'original' ```unnumbered``` links are based on loopback addresses: ipv4 and/or ipv6. There are also use cases for unnumbered links independent of loopback addresses.

My proposal is to resolve the ```unnumbered``` flag into ```{ 'ipv4': True/False, 'ipv6': True/False }``` early on in the code base, and get rid of the 'unnumbered' flag after that (also in the generated YAML)

I can pick this one up if you agree

**Comment by jbemmel** [comment]
None of the templates depend on the 'unnumbered' flag in the output (after my latest PR):
```
jeroen@jvm:~/srlinux/netlab$ grep -ri unnumbered netsim/ansible/*
netsim/ansible/templates/vrf/eos.ospfv2-vrf.j2: interface unnumbered hello mask tx 0.0.0.0
netsim/ansible/templates/ospf/eos.ospfv2.j2:{% if ospf.unnumbered is defined %}
netsim/ansible/templates/ospf/eos.ospfv2.j2: interface unnumbered hello mask tx 0.0.0.0
netsim/ansible/templates/bgp/cumulus_nvue.j2:                type: unnumbered
netsim/ansible/templates/bgp/srlinux.macro.j2:{# BGP unnumbered for IPv6 LLA, only supported for EBGP peers #}
netsim/ansible/templates/bgp/srlinux.macro.j2:{% set peer_group = "ebgp-unnumbered" + (('-' + n.local_as|string()) if n.local_as is defined else '') %}
netsim/ansible/templates/bgp/srlinux.macro.j2:   _annotate_receive-ipv4-check: "Allow IPv4 on IPv6 unnumbered interfaces"
netsim/ansible/templates/initial/junos.j2:        unnumbered-address lo0.0
netsim/ansible/templates/initial/nxos.j2: ip unnumbered loopback0
netsim/ansible/templates/initial/sros.gnmi.j2:    unnumbered:
netsim/ansible/templates/initial/sros.gnmi.j2:{# The interface name 'system' is special, also used for IP unnumbered #}
netsim/ansible/templates/initial/ios.j2: ip unnumbered {{ loopback_name|default('Loopback0') }} poll
netsim/ansible/templates/initial/eos.j2: ip address unnumbered {{ loopback_name|default('Loopback 0') }}
```

It only exists in the expected test output

**Comment by ipspace** [comment]
> My proposal is to resolve the `unnumbered` flag into `{ 'ipv4': True/False, 'ipv6': True/False }` early on in the code base

I think that's exactly how it's done:

https://github.com/ipspace/netlab/blob/dev/netsim/augment/links.py#L199

There's also dealing with various edge cases in

https://github.com/ipspace/netlab/blob/dev/netsim/augment/links.py#L267

> and get rid of the 'unnumbered' flag after that (also in the generated YAML)

We could do that in interface cleanup, for example after https://github.com/ipspace/netlab/blob/dev/netsim/augment/links.py#L598

> I can pick this one up if you agree

If you find it important, please go for it.

**Reply by jbemmel** [reply]
See https://github.com/ipspace/netlab/pull/435

**Comment by ipspace** [comment]
Implemented a long time ago
